src/sandstorm/sandstorm-ip-bridge.c++

```diff
@@ -0,0 +1,181 @@
+// Sandstorm - Personal Cloud Sandbox
+// Copyright (c) 2014 Sandstorm Development Group, Inc. and contributors
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This program is useful for including in Sandstorm application packages where
+// the application itself is a legacy HTTP web server that does not understand
+// how to speak the Cap'n Proto interface directly.  This program will start up
+// that server and then redirect incoming requests to it over standard HTTP on
+// the loopback network interface.
+
+#include "sandstorm/sandstorm-ip-bridge.h"
+
+#include <kj/debug.h>
+#include <netinet/ip.h>
+#include <linux/netfilter_ipv4.h>
+#include <unistd.h>
+
+namespace sandstorm {
+  namespace ipbridge {
+
+  class RefcountedAsyncIoStream: public kj::Refcounted {
+  public:
+    explicit RefcountedAsyncIoStream(kj::Own<kj::AsyncIoStream>&& stream)
+        : stream(kj::mv(stream)) {}
+    kj::Own<kj::AsyncIoStream> stream;
+  };
+
+  static const int BUFFER_SIZE = 8192;
+
+  // TCP handling
+  struct AcceptedConnection {
+    class Downstream final: public ByteStream::Server {
+    private:
+      kj::Own<RefcountedAsyncIoStream> connection;
+
+    public:
+      explicit Downstream(kj::Own<RefcountedAsyncIoStream> && connection) :
+        connection(kj::mv(connection)) {}
+
+      kj::Promise<void> write(WriteContext context) {
+        auto params = context.getParams();
+        return connection->stream->write(params.getData().begin(), params.getData().size());
+      }
+
+      kj::Promise<void> done(DoneContext context) {
+        connection->stream->shutdownWrite();
+        return kj::READY_NOW;
+      }
+    };
+
+  private:
+    kj::Own<RefcountedAsyncIoStream> connection;
+    capnp::byte buffer[BUFFER_SIZE];
+    TcpPort::Client port;
+    kj::Own<ByteStream::Client> upstream;
+
+  public:
+    explicit AcceptedConnection(kj::Own<kj::AsyncIoStream>&& connectionParam, TcpPort::Client port)
+        : connection(kj::refcounted<RefcountedAsyncIoStream>(kj::mv(connectionParam))), port(port) {
+      auto request = port.connectRequest();
+      request.setDownstream(kj::heap<Downstream>(kj::addRef(*connection)));
+      upstream = kj::heap<ByteStream::Client>(request.send().getUpstream());
+    }
+
+    kj::Promise<void> start() {
+      return messageLoop();
```

Comment:
Hmm, start() seems kind of redundant if it just calls another public function and returns its result.